numpy/ma/core.py

```diff
@@ -3171,7 +3164,7 @@ def __getitem__(self, indx):
                 dout._mask = _mask[indx]
                 # set shape to match that of data; this is needed for matrices
                 dout._mask.shape = dout.shape
-                dout._sharedmask = True
+                dout._sharedmask = np.may_share_memory(_mask, dout._mask)
```

Comment:
Tihs line could in principle change a result, if you do something crazy like:
```
arr = masked_arr[[0, 1, 2, 3]]  # not a view
view = arr[::2]
arr[4] = 5.  # will change view with change, triggers copy without change
view2 = arr[::4]  # view after
arr[6] = 6.  # would change view2, since copy was triggered.
```
so maybe could undo so that a warning is given. The logic is designed to avoid back propagation to the parent, since that creates copies at an arbitrary later time, it can do do weird things.